### PR TITLE
feat: Implement limits through stateful layer #59

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java [hunk 1]
```diff
@@ -119,11 +129,26 @@ public Future<?> handle(String deploymentId, String deploymentApi) {
...
                 .onSuccess(this::handleRequestBody)
                 .onFailure(this::handleRequestBodyError);
     }
 
+    private void handleRateLimitHit(RateLimitResult result) {
+        // Returning an error similar to the Azure format.
+        ErrorData rateLimitError = new ErrorData();
+        rateLimitError.getError().setCode(String.valueOf(result.status().getCode()));
+        rateLimitError.getError().setMessage(result.errorMessage());
+        log.error("Rate limit error {}. Key: {}. User sub: {}", result.errorMessage(), context.getProject(), context.getUserSub());
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: Trace/span id.

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 2]
```diff
@@ -79,11 +111,8 @@ private Limit getLimitByApiKey(ProxyContext context) {
         return role.getLimits().get(deployment.getName());
     }
 
-    private record Id(String key, String resource) {
-        @Override
-        public String toString() {
-            return String.format("key: %s, resource: %s", key, resource);
-        }
+    private String getRedisKey(String projectKey, String deploymentName) {
+        return String.format("limit.api.key.%s.%s", projectKey, deploymentName);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: In other places I use "prefix:" notation, e. g. "resource:", "lock:". Here I can suggest using rate_limit:

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 3]
```diff
@@ -52,18 +76,26 @@ public RateLimitResult limit(ProxyContext context) {
...
+                String json = bucket.get();
+                RateLimit rateLimit;
+                if (json == null) {
+                    return RateLimitResult.SUCCESS;
+                } else {
+                    rateLimit = ProxyUtil.MAPPER.readValue(json, RateLimit.class);
+                }
+                long timestamp = System.currentTimeMillis();
+                RateLimitResult result = rateLimit.update(timestamp, limit);
+                bucket.set(ProxyUtil.MAPPER.writeValueAsString(rateLimit));
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: Do you really need to set it back?

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 4]
```diff
@@ -5,41 +5,65 @@
...
 import lombok.extern.slf4j.Slf4j;
-
-import java.util.concurrent.ConcurrentHashMap;
+import org.redisson.api.RBucket;
+import org.redisson.api.RedissonClient;
+import org.redisson.client.codec.StringCodec;
 
 @Slf4j
+@RequiredArgsConstructor
 public class RateLimiter {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:23]: Have you thought about removing from cache?

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java (hunk 1 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: fixed

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: fixed

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: We change the state of the rate limit(e.g. counts). So yes, we have to set it back.

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 4 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:23]: set expiration for rate limit

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: You're right. We can update later if it's needed.

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java [hunk 5]
```diff
@@ -85,22 +85,29 @@ public Future<?> handle(String deploymentId, String deploymentApi) {
...
-            return respond(rateLimitResult.status(), rateLimitError);
+        Future<RateLimitResult> rateLimitResultFuture;
+        if (deployment instanceof Model) {
+            rateLimitResultFuture = proxy.getRateLimiter().limit(context);
+        } else {
+            rateLimitResultFuture = Future.succeededFuture(RateLimitResult.SUCCESS);
         }
 
+        return rateLimitResultFuture.onSuccess(result -> {
+            if (result.status() != HttpStatus.OK) {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: simplify by changing places: if ok then else

File: src/main/java/com/epam/aidial/core/limiter/RateLimit.java [hunk 6]
```diff
@@ -30,4 +32,8 @@ public synchronized RateLimitResult update(long timestamp, Limit limit) {
             return RateLimitResult.SUCCESS;
         }
     }
+
+    long timeout() {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimit.java]: Not needed?

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 7]
```diff
@@ -79,11 +128,8 @@ private Limit getLimitByApiKey(ProxyContext context) {
         return role.getLimits().get(deployment.getName());
     }
 
-    private record Id(String key, String resource) {
-        @Override
-        public String toString() {
-            return String.format("key: %s, resource: %s", key, resource);
-        }
+    private static String getPath(String deploymentName) {
+        return String.format("token/%s", deploymentName);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: let's use: deployment/tokens

File: src/main/java/com/epam/aidial/core/service/ResourceService.java [hunk 8]
```diff
@@ -208,6 +209,16 @@ public ResourceItemMetadata putResource(ResourceDescription descriptor, String b
         }
     }
 
+    public void computeResource(ResourceDescription descriptor, Function<String, String> fn) {
+        String redisKey = redisKey(descriptor);
+
+        try (var ignore = lockService.lock(redisKey)) {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/service/ResourceService.java:224]: The lock is not reetrant. Please use real server and service in the tests.

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 9]
```diff
@@ -5,65 +5,114 @@
...
+        }
         long timestamp = System.currentTimeMillis();
-        return rate.update(timestamp, limit);
+        return rateLimit.update(timestamp, limit);
+    }
+
+    private Void updateLimit(String path, ProxyContext context, long totalUsedTokens) {
+        String bucketLocation = BlobStorageUtil.buildUserBucket(context);
+        ResourceDescription resourceDescription = ResourceDescription.fromEncoded(ResourceType.LIMIT, bucketLocation, bucketLocation, path);
+        resourceService.computeResource(resourceDescription, new Function<String, String>() {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: why not to use lambda? You can create a method with sneaky throw.

File: src/main/java/com/epam/aidial/core/limiter/RateLimit.java (hunk 6 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimit.java]: right. will be removed

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 7 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: fixed

File: src/main/java/com/epam/aidial/core/service/ResourceService.java (hunk 8 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/service/ResourceService.java:224]: fixed

File: src/main/java/com/epam/aidial/core/limiter/RateLimit.java (hunk 6 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimit.java]: fixed

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 9 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java]: fixed

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java (hunk 5 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java]: fixed

File: src/test/java/com/epam/aidial/core/limiter/RateLimiterTest.java [hunk 10]
```diff
@@ -9,55 +9,65 @@
...
+import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
 public class RateLimiterTest {
 
+    @Mock
+    private Vertx vertx;
+
+    @Mock
+    private ResourceService resourceService;
```
Comment (artsiomkorzun) [inline src/test/java/com/epam/aidial/core/limiter/RateLimiterTest.java]: Please use the real service.

Comment (astsiapanay) [reply src/test/java/com/epam/aidial/core/limiter/RateLimiterTest.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Pull request merged**